rcl/src/rcl/security.c

```diff
@@ -0,0 +1,350 @@
+  const char * namespace_,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator
+);
+
+char * prefix_match_lookup(
+  const char * name,
+  const char * namespace_,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator
+);
+
+const security_lookup_fn_t g_security_lookup_fns[] = {
+  NULL,
+  exact_match_lookup,
+  prefix_match_lookup,
+};
+
+typedef enum ros_security_lookup_type_e
+{
+  ROS_SECURITY_LOOKUP_OVERRIDE = 0,
+  ROS_SECURITY_LOOKUP_MATCH_EXACT = 1,
+  ROS_SECURITY_LOOKUP_MATCH_PREFIX = 2,
+} ros_security_lookup_type_t;
+
+const char * const g_security_lookup_type_strings[] = {
+  "DIRECTORY_OVERRIDE",
+  "MATCH_EXACT",
+  "MATCH_PREFIX"
+};
+
+/// Return the directory that most closely matches a given name (longest-prefix match),
+/// scanning under base_dir.
+/**
+ * By using a prefix match, a name "my_name_123" will be able to load and use the
+ * directory "my_name" if no better match exists.
+ * \param[in] base_dir
+ * \param[in] name
+ * \param[out] matched_name must be a valid memory address allocated with at least
+ * _TINYDIR_FILENAME_MAX characters.
+ * \return true if a match was found
+ */
+static bool get_best_matching_directory(
+  const char * base_dir,
+  const char * name,
+  char * matched_name)
+{
+  size_t max_match_length = 0;
+  tinydir_dir dir;
+  if (NULL == base_dir || NULL == name || NULL == matched_name) {
+    return false;
+  }
+  if (-1 == tinydir_open(&dir, base_dir)) {
+    return false;
+  }
+  while (dir.has_next) {
+    tinydir_file file;
+    if (-1 == tinydir_readfile(&dir, &file)) {
+      goto cleanup;
+    }
+    if (file.is_dir) {
+      size_t matched_name_length = strnlen(file.name, sizeof(file.name) - 1);
+      if (0 ==
+        strncmp(
+          file.name, name, matched_name_length) &&
+        matched_name_length > max_match_length)
+      {
+        max_match_length = matched_name_length;
+        memcpy(matched_name, file.name, max_match_length);
+      }
+    }
+    if (-1 == tinydir_next(&dir)) {
+      goto cleanup;
+    }
+  }
+cleanup:
+  tinydir_close(&dir);
+  return max_match_length > 0;
+}
+
+char * exact_match_lookup(
+  const char * name,
+  const char * namespace_,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator)
+{
+  // Perform an exact match for the node/context's name in directory <root dir>/<namespace>.
+  char * secure_root = NULL;
+  // "/" case when root namespace is explicitly passed in
+  if (1 == strlen(namespace_)) {
+    secure_root = rcutils_join_path(ros_secure_root_env, name, *allocator);
+  } else {
+    char * fqn = NULL;
+    char * root_path = NULL;
+    // Combine namespace with name
+    // TODO(ros2team): remove the hard-coded value of the root namespace
+    fqn = rcutils_format_string(*allocator, "%s%s%s", namespace_, "/", name);
+    // Get native path, ignore the leading forward slash
+    // TODO(ros2team): remove the hard-coded length, use the length of the root namespace instead
```

Comment:
@ivanpauno meta: these TODOs seem to have been solved.

Reply:
I still see a "/" and a `+ 1` there, so I don't think they were solved.

Reply:
Oh, I see. Nevermind.